pkg/cmd/roachtest/tests/disk_stall.go

```diff
@@ -15,125 +15,441 @@ import (
 	"fmt"
 	"math/rand"
 	"runtime"
+	"strconv"
 	"strings"
 	"time"
 
 	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/option"
 	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/registry"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/spec"
 	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test"
 	"github.com/cockroachdb/cockroach/pkg/roachprod/install"
 	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
 )
 
 // registerDiskStalledDetection registers the disk stall test.
 func registerDiskStalledDetection(r registry.Registry) {
-	for _, affectsLogDir := range []bool{false, true} {
-		for _, affectsDataDir := range []bool{false, true} {
+	stallers := map[string]func(test.Test, cluster.Cluster) diskStaller{
+		"dmsetup": func(t test.Test, c cluster.Cluster) diskStaller { return &dmsetupDiskStaller{t: t, c: c} },
+		"cgroup/read-write/logs-too=false": func(t test.Test, c cluster.Cluster) diskStaller {
+			return &cgroupDiskStaller{t: t, c: c, readOrWrite: []string{"write", "read"}}
+		},
+		"cgroup/read-write/logs-too=true": func(t test.Test, c cluster.Cluster) diskStaller {
+			return &cgroupDiskStaller{t: t, c: c, readOrWrite: []string{"write", "read"}, logsToo: true}
```

Comment:
Is it worth testing all combinations of stallers and log/data devices? Or do we feel like a single staller can reliably test all failure modes?

Reply:
> Or do we feel like a single staller can reliably test all failure modes?

Not really, at the moment. Maybe down the line we can consolidate, but we want to be able to exercise exactly the test that a customer performs, regardless of how they perform it.